src/app/tabs/camera.cpp

```diff
@@ -1,135 +1,313 @@
+        this->cams_dropdown->setDisabled(true);
+        return false;
+    }
+
+    if (!this->local_device.isEmpty()) {
+        int pref_cam_index = this->cams_dropdown->findData(this->local_device);
+        if (pref_cam_index >= 0)
+            this->cams_dropdown->setCurrentIndex(pref_cam_index);
+    }
+    else {
+        QCameraInfo default_cam = QCameraInfo::defaultCamera();
+        if (!default_cam.isNull()) {
+            int default_cam_index = this->cams_dropdown->findData(default_cam.deviceName());
+            if (default_cam_index >= 0)
+                this->cams_dropdown->setCurrentIndex(default_cam_index);
+        }
+    }
+
+    this->cams_dropdown->setDisabled(false);
+    return true;
+}
+
+QWidget *CameraTab::input_widget(QRadioButton* network_radio)
+{
+    QWidget *widget = new QWidget(this);
+    QHBoxLayout *layout = new QHBoxLayout(widget);
+
+    QLineEdit *input = new QLineEdit(this->url, widget);
+    input->setContextMenuPolicy(Qt::NoContextMenu);
+    input->setFont(Theme::font_18);
+    input->setAlignment(Qt::AlignLeft);
+    connect(input, &QLineEdit::textEdited, [this](QString text) { this->url = text; });
+    connect(input, &QLineEdit::returnPressed, [this]() {
+        this->connect_network_stream();
+        this->config->set_cam_stream_url(this->url);
+    });
+
+    layout->addWidget(input, 4);
+
+    connect(input, &QLineEdit::textEdited, [=](){ network_radio->setChecked(true); });
+
+    return widget;
+}
+
+void CameraTab::update_network_status(QMediaPlayer::MediaStatus media_status)
+{
+    qInfo() << "camera status changed to: " << media_status;
+
+    switch (media_status) {
+        case QMediaPlayer::LoadingMedia:
+        case QMediaPlayer::LoadedMedia:
+        case QMediaPlayer::BufferedMedia:
+            this->status->setText("connecting...");
+            break;
+        default:
+            this->status->setText("connection failed");
+            emit disconnected();
+            break;
+    }
+}
+
+void CameraTab::connect_network_stream()
+{
+    connect(this->player, &QMediaPlayer::mediaStatusChanged,
+            [this](QMediaPlayer::MediaStatus media_status) { this->update_network_status(media_status); });
+    connect(this->player, QOverload<>::of(&QMediaPlayer::metaDataChanged), [this]() { emit connected_network(); });
+
+    QString pipeline = QString(
+                           "gst-pipeline: rtspsrc location=%1 ! decodebin ! video/x-raw ! videoconvert ! videoscale ! "
+                           "xvimagesink sync=false force-aspect-ratio=false name=\"qtvideosink\"")
+                           .arg(this->url);
+
+    qInfo() << "playing stream pipeline: " << pipeline;
+    this->player->setMedia(QUrl(pipeline));
+    this->player->play();
+}
+
+void CameraTab::connect_local_stream()
+{
+    if (this->local_cam != nullptr)
+        delete this->local_cam;
+
+    if (!this->local_cam_available(this->local_device)) {
```

Comment:
If you give QCamera a bogus device path it just connects to the default camera with no complaint. This code is an attempt to make behavior more predictable, particularly when connecting to a camera that was disconnected without the camera list dropdown being refreshed.